### ISSUE TITLE
If a job has a :pool option, the corresponding redis connection is serialized with that job

### ISSUE THREAD

**Issue by mnarayan01** [mperham/sidekiq#2655]
If a particular `Sidekiq::Worker` instance is given the `"pool"` option, the corresponding `Sidekiq::RedisConnection` will be serialized as the `"pool"` attribute for each queued job. This two consequences:

1.  A ton of memory is consumed as the serialized `Sidekiq::RedisConnection` is quite large.
2.  `Sidekiq::RedisConnection` is sometimes un-serializable (`Redis::Future` does not respond to `#respond_to?`).

My initial impression is that the pool is not needed, and simply gets thrown in via calls to [`Sidekiq::Client::normalized_hash`](https://github.com/mperham/sidekiq/blob/v3.5.3/lib/sidekiq/client.rb#L228-L235). This impression is enhanced by the fact that `Sidekiq::Middleware::Server::RetryJobs#attempt_retry` does not _appear_ to use it (ref https://github.com/mperham/sidekiq/blob/v3.5.3/lib/sidekiq/middleware/server/retry_jobs.rb#L128-L130).

**Comment by mperham** [comment]
Yes, IMO this option needs to go away.

**Comment by mperham** [comment]
I'm not able to reproduce your first claim.  The pool attribute looks like this to me: `"pool"=>"#<ConnectionPool:0x007fc832cd4d38>"`.  Not really huge by any means.

**Comment by mperham** [comment]
Can you tell me how to repro (2)?

**Comment by mnarayan01** [comment]
>  The pool attribute looks like this to me: `"pool"=>"#<ConnectionPool:0x007fc832cd4d38>"`.

It looks like the behavior I've observed is a result of [`active_support/core_ext/object/to_json`](https://github.com/rails/rails/blob/v4.0.0/activesupport/lib/active_support/core_ext/object/to_json.rb), which was deprecated in `activesupport 4.1` and removed in `4.2`. When pre-`activesupport 4.1`, the following:

```ruby
require "sidekiq"
require "sidekiq/api"

require "active_support/json/encoding"

QUEUE_NAME = "test_job"

Sidekiq::Queue.new(QUEUE_NAME).clear

class TestJob
  include Sidekiq::Worker

  sidekiq_options(:queue => QUEUE_NAME, :pool => Sidekiq.redis_pool)
end

TestJob.perform_async

Sidekiq.redis { |conn| puts conn.lrange("queue:#{QUEUE_NAME}", 0, 0).first }
```

yields (removing other keys):

```
"pool":{"size":5,"timeout":1,"available":{"create_block":{},"created":1,"que":[],"max":5,"mutex":{},"resource":{},"shutdown_block":null},"key":"current-2217807100"}
```

(this is when simply connecting to a minimal local redis; in other situations it can be _much_ larger).

Given that the (easiest) cause of the problematic behaviour is an old `activesupport`, I'm no longer sure whether this is something which needs to be addressed. (Locally I'm monkey-patching it out of `Sidekiq::Client#normalize_item`.)

> Can you tell me how to repro (2)?

It's semi-transient and seems to be based on congestion; I haven't looked that far into it. I'm guessing the above makes this irrelevant, but if not, I can look further.

**Comment by mnarayan01** [comment]
TLDR of the above: I would be fine with this issue being passed on.

For those using an older version of `activesupport` (or something else resulting in the same behavior), the following is a simply monkey-patch:

```ruby
Sidekiq::Client.prepend(
  Module.new do
    private def normalize_item(*)
      super.tap do |ret|
        ret.delete("pool")
      end
    end
  end
)
```

**Comment by mperham** [comment]
Thanks for clarifying.  Yeah, happy to let this one slide into obscurity.
